source/auth/auth.rst

```diff
@@ -453,10 +600,88 @@ Implementation
 Test Plan
 =========
 
-Tests have been defined in the associated files:
+Connection string tests have been defined in the associated files:
 
 * `Connection String <tests/connection-string.json>`_.
 
+---------------------------------------
+SCRAM-SHA-256 and mechanism negotiation
+---------------------------------------
+
+Testing SCRAM-SHA-256 requires server version 3.7.3 or later with
+``featureCompatibilityVersion`` of "4.0" or later.
+
+Drivers that allow specifying auth parameters in code as well as via
+connection string should test both for the test cases described below.
+
+Step 1
+------
+
+Create three test users, one with only SHA-1, one with only SHA-256 and one
+with both.  For example::
+
+    db.runCommand({createUser: 'sha1', pwd: 'sha1', roles: ['root'], mechanisms: ['SCRAM-SHA-1']})
+    db.runCommand({createUser: 'sha256', pwd: 'sha256', roles: ['root'], mechanisms: ['SCRAM-SHA-256']})
+    db.runCommand({createUser: 'both', pwd: 'both', roles: ['root'], mechanisms: ['SCRAM-SHA-1', 'SCRAM-SHA-256']})
+
+Step 2
+------
+
+For each test user, verify that you can connect and run a command requiring
+authentication for the following cases:
+
+- Explicitly specifying each mechanism the user supports.
+- Specifying no mechanism and relying on mechanism negotiation.
+
+For the example users above, the ``dbstats`` command could be used as a test
+command.
+
+For a test user supporting both SCRAM-SHA-1 and SCRAM-SHA-256, drivers should
+verify that negotation selects SCRAM-SHA-256.  This may require monkey
+patching, manual log analysis, etc.
+
+Step 3
+------
+
+For test users that support only one mechanism, verify that explictly specifying
+the other mechanism fails.
+
+For a non-existent username, verify that not specifying a mechanism when
+connecting fails with the same error type that would occur with a correct
+username but incorrect password or mechanism.  (Because negotiation with a
+non-existent user name causes an isMaster error, we want to verify this is
```

Comment:
The server responds with:

    MongoDB Enterprise > db.runCommand({isMaster : 1, saslSupportedMechs: "admin.foo"})
    {
	"ok" : 0,
	"errmsg" : "Could not find user foo@admin",
	"code" : 11,
	"codeName" : "UserNotFound"
    }

We should ensure that 11 is a stable error code that drivers can rely on, and add it to the spec as the error code to check for.

Reply:
I'll put it in the spec, but I've asked for confirmation.

Reply:
Please rephrase your response, as I'm not sure what you actually meant to say.

Reply:
Yeah, that was incoherent, wasn't it?!?

I've noted in the spec that the server will send error code 11.  I don't want to mandate in the *test* part of the spec that driver expose error code 11 so that drivers can continue to provide whatever exception type they already do for authentication errors, which may or may not include a server error code.

Better?

Reply:
A bit, yeah.  ;)

What I think we should mandate is that a driver should raise/throw its authentication-related exception type only when isMaster fails with error code 11, not if it fails for some unrelated reason.  Whether the error code itself it included in that exception depends on how that exception type is defined.

Reply:
Done. PTAL.